src/entry.rs

```diff
@@ -17,25 +21,43 @@ macro_rules! entry {
             panic!("Out of memory")
         }
 
+        #[cfg(not(target_arch = "riscv64"))]
         #[no_mangle]
         pub extern "C" fn _start() -> ! {
-            #[cfg(target_arch = "riscv64")]
-            unsafe {
-                asm!(
-                    "lw a0, 0(sp)",
-                    "addi a1, sp, 8",
-                    "li a2, 0",
-                    "call {}",
-                    "li a7, 93",
-                    "ecall",
-                    sym $main,
-                    options(noreturn)
-                );
-            }
-            #[cfg(not(target_arch = "riscv64"))]
-            panic!("This macro is only valid for riscv64 target");
+            panic!("ckb_std::entry is only valid for riscv64 target")
+        }
+
+        #[no_mangle]
+        unsafe extern "C" fn __ckb_std_main(
+            argc: core::ffi::c_int,
+            argv: *const *const core::ffi::c_char,
+        ) -> i8 {
+            let argv = core::slice::from_raw_parts(argv, argc as usize)
+                .iter()
+                .map(|arg| core::ffi::CStr::from_ptr(*arg))
+                .collect::<alloc::vec::Vec<_>>()
+                .leak();
```

Comment:
This function cost cycles, and I think most contracts don't act as a exec callee so argc/argv will not needed. Can use contract argv manually, and at the same time keep `main` a safe function.

Reply:
This can be optimized to just storing the pointer and length. No allocation, no looping. See the latest commit.

Reply:
Also this can be entirely optimized out if the args are never actually used and LTO is enabled.

Reply:
I have added `#[inline]` so LTO is not required.